src/Functions/bitSlice.cpp

```diff
@@ -0,0 +1,476 @@
+            length_byte = (length + offset_bit) / word_size;
+            over_bit = (length + offset_bit) % word_size;
+            if (over_bit && (length_byte || over_bit > offset_bit)) // begin and end are not in same byte OR there are gaps
+                length_byte++;
+        }
+
+        while (!src.isEnd())
+        {
+            size_t size = src.getElementSize();
+            if (length < 0)
+            {
+                length_byte = std::max(static_cast<ssize_t>(offset_byte) + (length / word_size), static_cast<ssize_t>(0));
+                over_bit = word_size + (length % word_size);
+                if (length_byte == 1 && over_bit <= offset_bit) // begin and end are in same byte AND there are no gaps
+                    length_byte = 0;
+            }
+            bool left_truncate = offset_byte > size;
+            bool right_truncate = length_byte > offset_byte;
+            size_t shift_bit = left_truncate ? 0 : offset_bit;
+            size_t abandon_last_bit = (over_bit && !right_truncate) ? word_size - over_bit : 0;
+            auto slice = src.getSliceFromRight(offset_byte, length_byte);
+            if (slice.size)
+                writeSliceWithLeftShift(slice, sink, shift_bit, abandon_last_bit);
+
+            sink.next();
+            src.next();
+        }
+    }
+
+    template <class Source>
+    void bitSliceDynamicOffsetBounded(Source && src, StringSink && sink, const IColumn & offset_column, const IColumn & length_column) const
+    {
+        const bool is_offset_null = offset_column.onlyNull();
+        const auto * offset_nullable = typeid_cast<const ColumnNullable *>(&offset_column);
+        const ColumnUInt8::Container * offset_null_map = offset_nullable ? &offset_nullable->getNullMapData() : nullptr;
+        const IColumn * offset_nested_column = offset_nullable ? &offset_nullable->getNestedColumn() : &offset_column;
+
+        const bool is_length_null = length_column.onlyNull();
+        const auto * length_nullable = typeid_cast<const ColumnNullable *>(&length_column);
+        const ColumnUInt8::Container * length_null_map = length_nullable ? &length_nullable->getNullMapData() : nullptr;
+        const IColumn * length_nested_column = length_nullable ? &length_nullable->getNestedColumn() : &length_column;
+
+
+        while (!src.isEnd())
+        {
+            size_t row_num = src.rowNum();
+            bool has_offset = !is_offset_null && !(offset_null_map && (*offset_null_map)[row_num]);
+            bool has_length = !is_length_null && !(length_null_map && (*length_null_map)[row_num]);
+            Int64 start = has_offset ? offset_nested_column->getInt(row_num) : 1;
+            Int64 length = has_length ? length_nested_column->getInt(row_num) : static_cast<Int64>(src.getElementSize());
+
+            if (start && length)
+            {
+                bool left_offset = start > 0;
+                size_t offset = left_offset ? start - 1 : -start;
+                size_t offset_byte;
```

Comment:
Can we initialize with some value here (e.g. with result of first branch)? It's dificult to keep that all wariables will be initalized in any branch.

Reply:
OK, I will change it.

Reply:
I think `if else` can clearly compare the differences between the two situations. I separate `offset` and `length`, now It looks clear.